tests.py

```diff
@@ -0,0 +1,303 @@
+    assert client.listdir('/foo2/') == ['bar/']
+    assert client.listdir('/foo2/bar/') == ['baz']
+    assert client.listdir('/foo3/bar/').sort() == ['baz1/', 'baz2/'].sort()
+
+
+@mkdirs_recursively(
+    tree={
+        'foo1': {
+            'bar': [],
+        },
+        'foo2': {
+            'bar': {
+                'baz1': [],
+                'baz2': None,
+            }
+        },
+        'foo3': {
+            'bar': {
+                'baz1': [],
+                'baz2': [],
+            }
+        }
+    }
+)
+def test_listdir_only_files(client):
+    assert client.listdir('/', only_files=True) == []
+    assert client.listdir('/foo1/', only_files=True) == []
+    assert client.listdir('/foo2/', only_files=True) == []
+    assert client.listdir('/foo2/bar/', only_files=True) == ['baz2']
+    assert client.listdir('/foo3/bar/', only_files=True) == []
+
+
+@mkdirs_recursively(
+    tree={
+        'foo': {
+            'bar': {
+                'baz': None
+            }
+        }
+    }
+)
+def test_isdir(client):
+    assert client.isdir('/foo/')
+    assert client.isdir('/foo/bar/')
+    assert not client.isdir('/foo/bar/baz')
+
+
+@mkdirs_recursively(
+    tree={
+        'foo1': {
+            'bar': [],
+        },
+        'foo2': {
+            'bar': {
+                'baz1': [],
+                'baz2': None,
+            }
+        },
+        'foo3': {
+            'bar': {
+                'baz1': [],
+                'baz2': [],
+            }
+        }
+    }
+)
+def test_rmdir(client):
+    client.rmdir('/foo1/bar/')
+    assert not client.exists('/foo1/bar/')
+
+    pytest.raises(ValueError, client.rmdir, '/foo2/bar/baz2')
+
+    pytest.raises(RuntimeError, client.rmdir, '/foo2/bar/')
+    client.rmdir('/foo2/bar/', recursive=True)
+    assert not client.exists('/foo2/bar/')
+
+    pytest.raises(RuntimeError, client.rmdir, '/foo3/')
+    client.rmdir('/foo3/', recursive=True)
+    assert not client.exists('/foo3/')
+
+    client.rmdir('/', recursive=True)
+    assert not client.exists('/foo1/')
+
+
+@mkdirs_recursively(
+    tree={
+        'foo1': {
+            'bar': [],
+        },
+        'foo2': {
+            'bar': {
+                'baz1': [],
+                'baz2': None,
+            }
+        },
+        'foo3': {
+            'bar': {
+                'baz1': [],
+                'baz2': [],
+            }
+        }
+    }
+)
+def test_rm(client):
+    client.rm('/foo1/bar/')
+    assert not client.exists('/foo1/bar/')
+
+    client.rm('/foo2/bar/baz2')
+    assert not client.exists('/foo1/bar/baz2')
+
+    pytest.raises(RuntimeError, client.rm, '/foo2/bar/')
+    client.rm('/foo2/bar/', recursive=True)
+    assert not client.exists('/foo2/bar/')
+
+    pytest.raises(RuntimeError, client.rm, '/foo3/')
+    client.rm('/foo3/', recursive=True)
+    assert not client.exists('/foo3/')
+
+    client.rm('/', recursive=True)
+    assert not client.exists('/foo1/')
```

Comment:
Maybe split to a couple of separate tests, so you can see the tree structure and the test code in a small amount of lines